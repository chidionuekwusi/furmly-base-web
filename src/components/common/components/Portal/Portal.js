import React from "react";
import styled, { ThemeProvider, withTheme } from "styled-components";
import PropTypes from "prop-types";
import ReactDOM from "react-dom";
import Overlay from "../Overlay";
import StyledButton from "../Button";

const ActionContainer = styled.div`
  align-self: flex-end;
`;
const ContentContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;

class Portal extends React.Component {
  constructor(props) {
    super(props);
    this.id = "furmly-dialog";
    this.portalElement = null;
    this.overlayClicked = this.overlayClicked.bind(this);
  }
  render() {
    return null;
  }
  overlayClicked() {
    if (this.props.overlayClicked) {
      this.props.overlayClicked();
    }
  }
  componentDidMount() {
    let id = this.props.portalId || this.id;
    let p = id && document.getElementById(id);
    if (!p) {
      p = document.createElement("div");
      p.addEventListener("click", this.clickOverlay);
      p.id = id;
      document.body.appendChild(p);
    }
    this.portalElement = p;
    this.componentDidUpdate();
  }
  componentWillUnmount() {
    this.portalElement.removeEventListener("click", this.overlayClicked);
    document.body.removeChild(this.portalElement);
  }
  componentDidUpdate() {
    ReactDOM.render(
      <ThemeProvider theme={this.props.theme}>
        <Overlay isOpen={this.props.isOpen}>
          <ContentContainer>
            {this.props.children}
            {this.props.actionButtons && (
              <ActionContainer>
                {this.props.actionButtons.map(x => (
                  <StyledButton onClick={x.onClick} key={x.key}>
                    {x.icon && <Icon icon={x.icon} />}
                    {x.content}
                  </StyledButton>
                ))}
              </ActionContainer>
            )}
          </ContentContainer>
        </Overlay>
      </ThemeProvider>,
      this.portalElement
    );
  }
}

Portal.propTypes = {
  theme: PropTypes.object.isRequired
};
export default Portal;
