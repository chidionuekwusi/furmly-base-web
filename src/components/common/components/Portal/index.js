import Portal from "./Portal";
import { withTheme } from "styled-components";

export default withTheme(Portal);
