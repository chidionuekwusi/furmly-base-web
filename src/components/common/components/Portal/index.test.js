import React from "react";
import { mount } from "enzyme";
import Portal from "./Portal";

const clickFn = jest.fn();
const fakeTheme = {};
describe("Portal", () => {
  it("should render correctly", () => {
    const component = mount(
      <Portal
        theme={fakeTheme}
        actionButtons={[
          {
            onClick: clickFn,
            content: "Click here please",
            key: "Click here my guy"
          }
        ]}
      >
        <span className="text">This is life</span>
      </Portal>
    );
    expect(component).toMatchSnapshot();
  });
});
