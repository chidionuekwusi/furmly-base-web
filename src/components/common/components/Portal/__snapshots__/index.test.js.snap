// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal should render correctly 1`] = `
<WithTheme(Portal)
  actionButtons={
    Array [
      Object {
        "content": "Click here please",
        "key": "Click here my guy",
        "onClick": [MockFunction],
      },
    ]
  }
  theme={Object {}}
>
  <Portal
    actionButtons={
      Array [
        Object {
          "content": "Click here please",
          "key": "Click here my guy",
          "onClick": [MockFunction],
        },
      ]
    }
    theme={Object {}}
  />
</WithTheme(Portal)>
`;
